lektor/packages.py

```diff
@@ -143,15 +150,37 @@ def install_local_package(package_root, path):
         dirs = os.listdir(tmp)
         if rv != 0 or len(dirs) != 1:
             raise RuntimeError("Failed to create egg info for local package.")
-        requires = os.path.join(tmp, dirs[0], "requires.txt")
+        requires_path = os.path.join(tmp, dirs[0], "requires.txt")
+
+        if os.path.isfile(requires_path):
+            # We have dependencies, install them!
+            requirements_path = requiriements_txt_from_requires_file_in_same_directory(
```

Comment:
Spelling aside, I wonder whether this long function name is really helping readability.

(The "in the same directory", I think, is an implementation detail and can be omitted from the function name with no loss of value.  The caller of this function really has no reason to care where the requirements.txt file is created.)

The essence of what this function does is: extract the `install_requires` from the egg-info `requires.txt`, and write them to `requirements.txt` suitable for passing to pip.

Note that we are lucky — and have been taking advantage of the fact — that the format of setuptool's `requires.txt` is similar to that of pip’s `requirements.txt` as it is.  Really, their formats are defined by two different specifications for two different (though related) purposes.  `Requires.txt` is part of the egg-info metadata — its format is specified by setuptools (which is who defined the egg format). `Requirements.txt` is really mostly a way of passing arguments to pip — its format is specified by pip.

The reason the `requires.txt` was used as a `requirements.txt` in the first place was that is was easy.  Now that it's not quite as easy, maybe the use of a `requirements.txt` is not needed at all — just pass the requirements we'd like installed to pip as command-line arguments.

It might be clearer to refactor a bit higher into `install_local_package` to make it clearer what is being done.  Perhaps:

```python
def install_local_package(package_root, path):
    """This installs a local dependency of a package."""
    # XXX: windows
    env = dict(os.environ)
    env["PYTHONPATH"] = package_root

    # Step 1: generate egg info and link us into the target folder.
    [...] (unchanged)

    # Step 2: install additional requirements
    for requirement in get_package_install_requires(path):
        # NB: this bastardizes the `package` argument of `download_and_install_package`
        # as we're passing a *requirement specifier* rather than just a package name.
        # I think it will work, but might want to adjust the signature of `download_and_install_package`
        # to provide a way to make this more explicit.  (While at it, perhaps provide a way to provide
        # multiple requirement specifiers in one go?)
        download_and_install_package(package_root, requirement)
```

The `get_package_install_requires` (probably) runs `setup.py egg_info` (as was done previously by `install_local_package`) and returns a sequence containing the initial lines — up to the first section marker ­of the resulting `requires.txt`.

Okay, I've rambled now, but, I think, have arrived at a worthwhile place:

The only reason `setup.py egg_info` was being run in `install_local_package` is to get at the packages `install_requires`.  Abstracting that whole process to a new function makes that logic clear.  Also, it abstracts implementation details.

Reply:
I've chosen this way of implementing this requires.txt -> requirements.txt cleanup step in a way that tries to be minimally invasive to the existing source code.

That being said, I too think that extracting the whole step including the `egg_info` generation would be much cleaner. As I now know that you're willing to take that change, I'll see what I can come up with.

Reply:
@dairiki I've pushed a change that tries to get the spirit of what you said - what do you think about it?